### DISCUSSION TITLE
24. 멱등성을 보장하는 HTTP 메서드에 대해 설명해 주세요.

### DISCUSSION THREAD

**Discussion by eun-byeol** [jewelry-box/cs-interview#25] [General]
멱등성을 보장하는 HTTP 메서드에 대해 설명해 주세요.

**Comment by linirini** [comment]
멱등성은 **같은 요청을 여러 번 반복해서 보내더라도 결과가 달라지지 않는 성질**을 의미한다.
### 멱등성을 보장하는 HTTP 메서드
1. GET
- 서버에서 데이터를 조회하는 데 사용
- 특징
  - 데이터를 읽기만 하므로, 여러 번 요청해도 서버의 상태에 영향을 주지 않는다.
  - 동일한 리소스에 대해 여러 번 GET 요청을 보내더라도 결과는 동일한 데이터를 반환한다.
2. HEAD
- GET과 유사하지만, 응답에서 헤더 정보만 반환하고 본문은 포함하지 않는다.
- 특징
  - 서버는 데이터를 조회하기만 하므로, 멱등서을 보장한다.
  - 동일한 HEAD 요청 여러번에도 서버 상태는 변하지 않는다.
3. PUT
- 서버에 리소스를 생성하거나 업데이트를 할 때 사용된다.
- 특징
  - 동일한 리소스에 대해 동일한 데이터를 여러 번 PUT 요청하면, 결과는 항상 동일하게 유지된다.
  - 기존 데이터를 덮어쓰기 때문에 결과는 동일하다.
4. DELETE
- 서버에서 리소스를 삭제하는데 사용된다.
- 특징
  - 동일한 DELETE 요청을 여러 번 보내더라도 결과는 동일하다.
  - 첫 요청에 리소스가 삭제되고, 이후 요청에 리소스가 이미 삭제되어 있어 더 이상 변화가 없다.
5. OPTIONS
- 서버가 특정 리소스에서 허용하는 메서드 목록을 반환한다.
- 특징
  - 여러 번 OPTIONS 요청을 보내더라도 동일한 허용 메서드 목록을 반환하므로 멱등성을 보장한다.
### HTTP 멱등성의 중요성
1. 안정성
  멱등성을 보장하는 메서드는 클라이언트가 네트워크 오류로 인해 동일한 요청을 반복하더라도 안정적으로 작동한다.
2. 캐싱과 최적화
  GET 같은 멱등성 보장 요청은 캐싱이 가능하므로 성능 최적화에 유리하다.
### 참고

![image](https://github.com/user-attachments/assets/f11586a9-d3a2-4ec8-bd7a-b9de0c48ef0f)
